databricks/sdk/core.py

```diff
@@ -849,6 +852,30 @@ def __init__(self,
         self.kwargs = kwargs
 
 
+class RateLimiter:
+    def __init__(self, max_requests: int):
+        if max_requests is None:
+            max_requests = 30
+        self._max_requests = int(max_requests)
+        self._lock = threading.Lock()
+        self._requests = collections.deque()
+
+    def _window_size(self):
+        return self._requests[-1] - self._requests[0]
+
+    def throttle(self):
+        with self._lock:
+            burst_period_seconds = 1
+            if len(self._requests) >= self._max_requests:
+                sleep = self._window_size()
+                logger.debug(f'Request throttled. Sleeping for {sleep}s')
+                time.sleep(sleep)
+            self._requests.append(time.time())
+            while self._window_size() >= burst_period_seconds:
+                self._requests.popleft()
+            return self
```

Comment:
Maybe I'm misunderstanding the algorithm, but what makes sense to me is:
```
start_time = time.time()
while self._requests[0] < start_time - burst_period_seconds
  self._requests.popleft()
if len(self._requests) > self._max_requests:
  sleep_time = burst_time_seconds - self._requests[0]
  time.sleep(sleep_time)
self._requests.append(time.time())
```
Suppose that 30 requests are made within a second. In the current implementation, the next request will always sleep for burst_period_seconds, even if made more than one second later. This version would clear the request queue before checking if sleeping is needed.